### PR TITLE
Overpass API import script

### PR DESCRIPTION
**Pull request by mvaaltola**
Add script for importing data from Overpass API to PostGIS.

todos:

- [x] use this instead of import_osm.sh
- [x] remove unneeded osm2pgsql dependency from image
- [x] merge conflict in import_flickr
- [x] pin test/dev dependencies using pip-tools (in #34)
- [x] separate tests to test directory (wontfix. larger refactoring needed)
- [x] run tests automatically in GH (in #24)

### CODE REVIEW

File: requirements-dev.in [hunk 1]
```diff
@@ -0,0 +1,4 @@
+psycopg2
```
Comment (Rikuoja) [inline requirements-dev.in:1]: Why is this needed instead of psycopg2-binary? 🤔

File: scripts/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  geoviz_test_db:
+    image: postgis/postgis:latest
+    container_name: geoviz_test_db
+    volumes:
+      - ./pgdata:/var/lib/postgresql/data/pgdata
+    ports:
+      - "5400:5432"
+    environment:
+      - PGDATA=/var/lib/postgresql/data/pgdata
+      - POSTGRES_PASSWORD=postgres
```
Comment (Rikuoja) [inline scripts/docker-compose.yml:13]: Should this file be combined with root docker-compose.yml, or do you think we should have a separate docker-compose.yml for tests? Perhaps put this on root level or create a /test directory for this?

File: scripts/import_osm.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        """Parse input arguments to a dict."""
+        slug = args["city"]
+        if not slug:
+            raise AssertionError("You must specify the city name.")
+        # BBOX (minx, miny, maxx, maxy) takes precedence over city
+        arg_bbox = args["bbox"]
+        if arg_bbox:
+            minx, miny, maxx, maxy = map(float, arg_bbox.split(", "))
+            city_name = ""
+        else:
```
Comment (Rikuoja) [inline scripts/import_osm.py:38]: Perhaps we don't need these anymore. Probably nobody uses these slugs anyway, so better get rid of useless stuff, they are old leftovers in the accessibility importer too.

File: scripts/import_osm.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                city_name = "Copenhagen, Denmark"
+                minx, miny, maxx, maxy = 12.42000, 55.61000, 12.65000, 55.78000
+            elif slug == "HEL":
+                city_name = "Helsinki, Finland"
+                minx, miny, maxx, maxy = 24.82345, 60.14084, 25.06404, 60.29496
+            elif slug == "WAR":
+                city_name = "Warsaw, Poland"
+                minx, miny, maxx, maxy = 20.79057, 52.09901, 21.31300, 52.38502
+            else:
+                raise AssertionError("You must specify a bounding box or a known city.")
```
Comment (Rikuoja) [inline scripts/import_osm.py:50]: Actually here we should always require boundingbox. The import.py script does geocoding for city names, so I don't think any importers should care about city names anymore, unless they specifically *need* a city name for some reason.

File: scripts/import_osm.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+sys.path.insert(0, "..")
+from models import OSMPoint
+from osm_tags import tags_to_filter
+
+
+class OsmImporter(object):
+    """Import OSM data to PostGIS using Overpass API."""
+
+    def __init__(self, args: Dict):
+        args = self._parse_args(args)
```
Comment (Rikuoja) [inline scripts/import_osm.py:21]: Good idea to separate the parsing method. Probably should unify this for all importers in the future, because all the importers do quite the same thing.

File: scripts/import_osm.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+        pois = pois.rename(columns={"osmid": "node_id"})
+        pois = pois.rename(columns={"geometry": "geom"})
+        pois = pois.set_index("node_id")
+        pois["geom"] = pois["geom"].apply(lambda geom: WKTElement(geom.wkt, srid=4326))
+
+        pois["tags"] = pois[tag_columns].apply(lambda x: x.to_json(), axis=1)
+        pois = pois.drop(tag_columns, axis=1)
+
+        pois.to_sql(name=OSMPoint.__tablename__, con=self._engine, schema=self.slug, if_exists="append",
```
Comment (Rikuoja) [inline scripts/import_osm.py:83]: Probably "append" is not needed if we always drop the table beforehand? Doesn't do harm either, though.

File: scripts/test_import_osm.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    assert session.query(OSMPoint).first() is None
+    session.close()
+
+
+def test_data_loading(db_engine, helsinki_importer):
+    helsinki_importer.run()
+
+    with db_engine.connect() as con:
+        cur = con.execute("""select tags->'name', tags->'addr:city' from "Helsinki".osmpoints o where node_id = 25038585""")
+        assert cur.fetchall() == [("Polyteekkarimuseo", "Espoo")]
```
Comment (Rikuoja) [inline scripts/test_import_osm.py:58]: This is essential.

File: scripts/import_osm.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        self._engine.execute(CreateSchema(self.slug))
+
+        OSMPoint.__table__.drop(self._engine, checkfirst=True)
+        OSMPoint.__table__.create(self._engine)
+
+    def _get_amenities(self):
+        try:
+            (minx, miny, maxx, maxy) = self.bbox
+            return ox.geometries.geometries_from_bbox(maxy, miny, maxx, minx, tags=tags_to_filter)
+        except (TypeError, ValueError):
```
Comment (Rikuoja) [inline scripts/import_osm.py:99]: Probably we don't need this? Seems the importer will not run without providing a bounding box anyway (apart from the three old slugs HEL, COP, WAR that we should deprecate)?

Or alternatively, if we want to keep the ox geocoding from city name, AssertionError in _parse_args should only check that *either* bbox *or* city is set. But simpler would be to just require bbox.

File: scripts/import_osm.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        pois["tags"] = pois[tag_columns].apply(lambda x: x.to_json(), axis=1)
+        pois = pois.drop(tag_columns, axis=1)
+
+        pois.to_sql(name=OSMPoint.__tablename__, con=self._engine, schema=self.slug, if_exists="append",
+                    dtype={"geom": Geometry(geometry_type="POINT", srid=4326)})
+
+    def _initialise_db(self) -> None:
+        """Initialises OSM points table and returns a new DB Session."""
+        with self._engine.connect() as con:
+            con.execute("CREATE EXTENSION postgis")
```
Comment (Rikuoja) [inline scripts/import_osm.py:89]: Here we should add `try... except psycopg2.errors.DuplicateObject`. This will crash if the database exists and has the extension already.

File: scripts/import_osm.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        tag_columns = list(pois.columns)
+        tag_columns.remove("osmid")
+        tag_columns.remove("geometry")
+
+        pois = pois.rename(columns={"osmid": "node_id"})
+        pois = pois.rename(columns={"geometry": "geom"})
+        pois = pois.set_index("node_id")
+        pois["geom"] = pois["geom"].apply(lambda geom: WKTElement(geom.wkt, srid=4326))
+
+        pois["tags"] = pois[tag_columns].apply(lambda x: x.to_json(), axis=1)
```
Comment (Rikuoja) [inline scripts/import_osm.py:80]: This one results in a huge json object of tags with all possible json values. I don't think we want all these strings with nulls in the database, it becomes huge and unreadable 😢 
```
{"bar": null, "dog": null, "ele": null, "fax": null, "fee": null, "gay": null, "hgv": null, "lit": null, "mtb": null, "pub": null, "ref": null, "url": null, "area": null, "beer": null, "cafe":
 null, "club": null, "cold": null, "date": null, "dogs": null, "fish": null, "food": null, "foot": null, "heat": null, "male": null, "meat": null, "name": "Polyteekkarimuseo", "note": null, "roof": null, 
"room": null, "ship": null, "shop": null, "type": null, "ways": null, "wifi": null, "FIXME": null, "asian": null, "blind": null, "books": null, "brand": null, "craft": null, "depth": null, "email": null, 
"fixme": null, "floor": null, "games": null, "halal": null, "hoops": null, "horse": null, "image": null, "is_in": null, "layer": null, "level": null, "lgbtq": null, "lunch": null, "music": null, "nodes": 
null, "parts": null, "phone": null, "place": null, "power": null, "sauna": null, "sells": null, "shoes": null, "sport": null, "stars": null, "table": null, "taxon": null, "topic": null, "trade": null, "ac
cess": null, "bakery": null, "beauty": null, "boules": null, "branch": null, "brunch": null, "charge": null, "colour": null, "erotic": null, "female": null, "fenced": null, "hazard": null, "height": null,
 "indoor": null, "length": null, "mobile": null, "museum": null, "noname": null, "office": null, "opened": null, "origin": null, "public": null, "rental": null, "repair": null, "robbed": null, "screen": n
ull, "shower": null, "source": null, "toilet": null, "unisex": null, "CHECKME": null, "address": null, "alcohol": null, "amenity": null, "archery": null, "barrier": null, "bicycle": null, "brewery": null,
 "butcher": null, "chicken": null, "clothes": null, "covered": null, "cuisine": null, "culture": null, "date_on": null, "disused": null, "ele:msl": null, "highway": null, "karaoke": null, "landuse": null,
 "leisure": null, "massage": null, "max_age": null, "min_age": null, "name:ar": null, "name:ca": null, "name:da": null, "name:de": null, "name:en": null, "name:es": null, "name:et": null, "name:fa": null,
 "name:fi": null, "name:fr": null, "name:id": null, "name:it": null, "name:ja": null, "name:ko": null, "name:nl": null, "name:nn": null, "name:no": null, "name:ro": null, "name:ru": null, "name:se": null,
 "name:sv": null, "name:uk": null, "name:zh": null, "natural": null, "organic": null, "outdoor": null, "parking": null, "phone_1": null, "portier": null, "rooftop": null, "sensory": null, "service": null,
 "shelter": null, "smoking": null, "surface": null, "terrace": null, "toilets": null, "tourism": "museum", "twitter": null, "vending": null, "website": "https://www.polyteekkarimuseo.fi/", "alt_name": nul
l, "boundary": null, "building": null, "capacity": null, "currency": null, "delivery": null, "diet:raw": null, "drive_in": null, "end_date": null, "entrance": null, "facebook": null, "fuel:cng": null, "fu
el:e10": null, "fuel:e85": null, "fuel:lng": null, "fuel:lpg": null, "gasoline": null, "historic": null, "identity": null, "int_name": "Polytechnical Museum", "loc_name": null, "location": null, "man_made": null, "material": null, "memorial": null, "motorcar": null, "name:old": null, "old_name": null, "operator": null, "provides": null, "religion": null, "seasonal": null, "shooting": null, "takeaway": null, "veikkaus": null, "was:name": null, "was:shop": null, "waterway": null, "wikidata": "Q11888594", "addr:city": "Espoo", "addr:door": null, "addr:full": null, "addr:name": null, "addr:unit": null, "amenity_1": null, "architect": null, "athletics": null, "automated": null, "bath:type": null, "breakfast": null, "cinema:3D": null, "cocktails": null, "cuisine_1": null, "cuisine_2": null, "diet:meat": null, "ele:wgs84": null, "fee:adult": null, "fee:child": null, "fireplace": null, "fuel:MPÖ": null, "full_name": null, "furniture": null, "glass:new": null, "high_seat": null, "highchair": null, "ice_cream": null, "leaf_type": null, "leisure_1": null, "max_level": null, "min_level": null, "ref:vatin": null, "ship:type": null, "substance": null, "wholesale": null, "wikipedia": "fi:Polyteekkarimuseo", "access:dog": null, "addr:floor": null, "addr:place": null, "beer_brand": null, "car_repair": null, "created_by": null, "diet:halal": null, "diet:vegan": null, "dispensing": null, "drink:wine": null, "ele:source": null, "embankment": null, "fair_trade": null, "fee:family": null, "fence_type": null, "healthcare": null, "high_chair": null, "lunch:menu": null, "lunch:warm": null, "maintained": null, "piste:type": null, "playground": null, "restaurant": null, "roof:shape": null, "sauna:fuel": null, "short_name": null, "start_date": null,
```

Comment (Rikuoja) [reply scripts/import_osm.py:80]: I think at least filter out *all* the null values.

Or, even better, since we are only interested in amenities, import `tags_to_filter` from `osm_tags.py` and *only* save those, and filter out all the nulls.

Otherwise, the tag filter doesn't work correctly I'm afraid, since it checks the presence of OSM tags, and to_json() seems to add
*every* tag to *every* object.

File: scripts/import_osm.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        pois = pois.drop(tag_columns, axis=1)
+
```
Comment (Rikuoja) [inline scripts/import_osm.py:82]: I think we should add here some basic logging so the user running the script sees if anything has been imported. Just some lines like in the other importers like
`self.logger.info(f"Found {len(walk_access_dict)} accessibility nodes, importing...")` and
`self.logger.info(f"Saving {len(nodes_to_save)} accessibility nodes...")`

**Review by Rikuoja — CHANGES_REQUESTED** [review]
Looks excellent! Added some comments and things to fix before merging.

The details about file paths, testing etc. can be fixed also after merging.

File: scripts/import_osm.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        args = self._parse_args(args)
+        self.slug = args["slug"]
+        self.bbox = args["bbox"]
+        self.city = args["city_name"]
+        self._engine = self._create_engine()
+
+    @staticmethod
+    def _parse_args(args) -> Dict:
+        """Parse input arguments to a dict."""
+        slug = args["city"]
```
Comment (Rikuoja) [inline scripts/import_osm.py:30]: Ah, one more thing. In the database, the slug should always be lowercase, so I think we should use slugify here, or perhaps add argument called `slug` like in the other importers.

Actually `slug` argument is probably needed, because import.py calls also other importers by slug. This way, all the imported datasets will end up in the same schema with lower-case slug.

File: scripts/import_osm.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        pois = pois.drop(tag_columns, axis=1)
+
+        pois.to_sql(name=OSMPoint.__tablename__, con=self._engine, schema=self.slug, if_exists="append",
+                    dtype={"geom": Geometry(geometry_type="POINT", srid=4326)})
+
+    def _initialise_db(self) -> None:
+        """Initialises OSM points table and returns a new DB Session."""
+        with self._engine.connect() as con:
+            con.execute("CREATE EXTENSION postgis")
+        self._engine.execute(CreateSchema(self.slug))
```
Comment (Rikuoja) [inline scripts/import_osm.py:90]: Same here, this will crash if the schema exists.

File: import.py [hunk 14]
```diff
@@ -191,23 +162,13 @@ def mark_imported(dataset: str):
...
 if "osm" in datasets:
     logger.info(f"--- Importing OSM data for {city} ---")
-    import_path = os.path.join(os.path.dirname(__loader__.path), "scripts", "import_osm.sh")
-    for continent in CONTINENTS:
-        # Nominatim does not provide us with the continent. Will have to do some guessing
-        if not os.system(f"{import_path} {continent} {country} {slug} {osm_extracts_api_key}"):
-            # success!
-            mark_imported("osm")
-            break
+    osm_importer = OsmImporter({"slug": slug, "city": city, "bbox": ", ".join([str(coord) for coord in bbox]), "logger": logger})
```
Comment (Rikuoja) [inline import.py:169]: Here, we have to call the importer a bit different from the other importers below. Perhaps change the function signature so all importers use the same parameters?

File: scripts/import_osm.py (hunk 12 above)
Comment (Rikuoja) [reply scripts/import_osm.py:30]: Actually misread this. This slug only means the old-fashioned three-letter slug we don't need. Just get rid of three-letter slugs.

File: scripts/import_osm.py (hunk 10 above)
Comment (mvaaltola) [reply scripts/import_osm.py:72]: Will fix this later once I get the testing environment to work again...

File: scripts/import_osm.py (hunk 9 above)
Comment (mvaaltola) [reply scripts/import_osm.py]: Added IF NOT EXISTS

File: scripts/import_osm.py (hunk 10 above)
Comment (mvaaltola) [reply scripts/import_osm.py]: Should be fixed now.

File: util.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
+import logging
+import sys
+from pathlib import Path
+
+LOG_PATH = Path('/app/logs')
```
Comment (Rikuoja) [inline util.py:5]: I'm afraid we will have to use
```
log_file = os.path.join(os.path.dirname(__loader__.path), IMPORT_LOG_PATH, f"{city}.log")
```

or something similar here, so that this setup works
1) inside the container
2) outside the container, with a totally different directory inside the repo,
3) when called on the command line in the top directory
4) when called by flask.

**Review by Rikuoja — CHANGES_REQUESTED** [review]
Log path needs fixing, but otherwise this looks very good!

File: util.py (hunk 15 above)
Comment (Rikuoja) [reply util.py:5]: Otherwise I will get
```
FileNotFoundError: [Errno 2] No such file or directory: '/app/logs/helsinki.log'
```
when running locally.

**Review by Rikuoja — APPROVED** [review]
